### ISSUE TITLE
API documentation is garbage [Issue]

### ISSUE THREAD

**Issue by aspenforest** [iexg/IEX-API#993]
**Clearly describe the problem**
Incorrectly documented endpoints, missing or incorrect path parameter descriptions, missing or incorrect query string parameters, inadequate examples of available methods, etc. 

**What time did the problem occur? (with timezone)**
It's continuous.

**What is the expected behavior?**
Describe how the API endpoints actually work.

**Your environment (e.g. browser, OS, API version, language):**
Human brain, eyes, english

**API URL (if applicable)**
The whole https://iexcloud.io/docs

**What are the steps to reproduce the bug?**
RTFM

**Additional context**

**Comment by evancarter-iex** [comment]
@aspenforest Hi there, do you have specific examples?

**Comment by lars670** [comment]
:-)

**Comment by aspenforest** [comment]
Balance Sheet, Dividends, Cash Flow, ... I haven't yet read further, but bet your ass I expect the rest to be similar disaster.

**Comment by aspenforest** [comment]
In one case, "range" is a path parameter, in another case its a query string parameter. 
What would have to go through ones head to come to the conclusion that implementing the same feature in two different ways is a good idea

**Comment by lars670** [comment]
On the top of my head - this morning (EU time) I noticed that the new end-points for financial reports did not work as per the example in the Doc, and the link to ISIN conversion was dead.

**Comment by aspenforest** [comment]
**For example, Balance Sheet:**

_GET /stock/{symbol}/balance-sheet/{last}/{field}_ 

the {field} path parameter is nowhere documented in that method. 
Query string parameters section contains description of parameter "last", which is 
absent from available methods description

**Comment by aspenforest** [comment]
**Example #2: Batch requests**

_GET /stock/{symbol}/batch_

Examples section contains an api call with "last" query param, which is not documented in that api call method. The "type" query param is describe as being limited to 10 types. The 10 available types are nowhere documented.

**Comment by aspenforest** [comment]
If I were to continue, it would take me until tomorrow, as i've only scratched the surface of the available api calls. If the beginning is that bad, I'm not expecting anything better from the rest of it. 
If anything, I would suggest IEX to stop outsourcing documentation writing to indian tech support.

**Comment by evancarter-iex** [comment]
@aspenforest thank you we will fix the issues you mentioned and will look into if we see other problems.

**Comment by evancarter-iex** [comment]
@aspenforest Those are fixed along with @lars670 linking issue. Let me know if you see anything else as you go. And thanks for taking time to list a few of them out.

**Comment by evancarter-iex** [comment]
I am going to close this. If you come across additional examples, please report them so we can update the docs.

**Comment by joshuablackburn-iex** [comment]
@aspenforest  We provide this GitHub community as a transparent way to communicate with our users, and to create a public reference for fixes and enhancements. While we appreciate every issue raised in this forum, we do expect a level of professionalism. Some of the issues you've raised need to be addressed in the docs. Others are there for a reason. We are not here to create the exact perfect implementation of REST, for which developers disagree regularly, but an API that is easy to use.

If you have such a major issue with our system or docs, you are welcome to email us directly. This is a commercial product and we will take appropriate steps to maintain professionalism in this forum.

**Comment by jjreego** [comment]
Only snippets of the  IEX Cloud API are defined. It is difficult to run the examples as there are only snippets. Most of the code snippets do not specify the operating system, development environment or dependencies required to execute the snippets of code. Flow control, interactions between messages, and internal state machines for the sessions, are not defined. I spent over 40 hours trying to get the snippets of code running under Windows 10 with Visual Studio 2019. I also tried to run the Java Script examples using the latest Node.js, Oracle JVM, and Eclipse. The third part API libraries often assert and are undocumented as well.  I am more than happy to test things out - but there is very little to go on. I had to write a parser in C# just to decode what the IEX Cloud was sending back to me. I should not have to write all this support code.

**Comment by schardtbc** [comment]
https://cloud.iexapis.com/stable/stock/AAPL/quote?token=pk... 

where pk_... is your key from your iexcloud.io account.
Enter into any web browser and get the results. How is this difficult, or obscure. Results are standard JSON (javascript object notation).

**Comment by jjreego** [comment]
A web browser interface is not useful for what I want to do. I agree that it is simple - just not useful for what I want to do.  I have looked at all  the 3rd party projects that support the IEX Cloud as well as the pre-cloud interface. There is  quite a bit of code needed to query the IEX Cloud and parse the results.  The third parties have implemented synchronous, asynchronous, and streaming interfaces.  Every implementation was different - and buggy. Only some of the queries were supported. Many of the interfaces worked during the day but crashed during non-trading hours due to null values being returned. And most of the projects have not been updated for several months. All of the projects have severe bugs that make the code unusable. In some cases there is only a .DLL with no documentation on how to use the .DLL. If you look at the entire IEX Cloud interface it is not trivial.  It is inefficient for every customer to build their own API. Not everyone is a JSON or REST expert.  And even less people understand the complexity of the IEX Cloud interface.  I just want an API library so I can effectively access the data.

**Comment by jjreego** [comment]
Bruce, I did like the iexcloud_api_wrapper. I did get it to work under Visual Studio 2019.  I am not a Java Script expert.  The asynchronous interface made things more difficult. About all I could do is write the results to the console.  Sometimes I received multiple events, which were not expected. Not being a JS expert I was not able to make queries where the interface was different.  I tried looking at the usage file - but could not figure out what the corresponding JS code should look like.  JS is not a good choice for what I want to do.  I think a few more examples where the interface has changed would be helpful.

**Comment by schardtbc** [comment]
What specifically are you trying to do? What language do you prefer? C#?, Java?

I agree it is a complex API, like most REST API's. Again, Like most REST API's, it is not 100% consistent in its design. The Documentation is on-par with vast majority of API documentation on the web: provide the REST URL for each endpoint, give a sample of the JSON data structure returned. 

For a language like C# coverage of the complete API is a large amount of code, as each endpoint requires a DTO class (data transfer object) and code to deserialize the Api response for each endpoint and that is only the beginning of any real application development.

My suggestion to you would be to contact the author of the IEXCloudDotNet wrapper and see if you can work with him on a solution that meets your needs.

While it would be quite easy to make a synchronous javascript api for iexcloud, that would not fit my use case.

**Comment by jjreego** [comment]
I had decent success with your iexcloud_api_wrapper.  I used the
stock_quote and ref_data functions. Then yesterday I started getting a
ETIMEDOUT message. I used to be able to process about 100 messages per
second. Now I get about 2 messages per second along with the ETIMEDOUT
message which occurs randomly from right away to several minutes
later. I also changed the code to wait for the response before
requesting a new quote. So there is only one active request at a time.

Here is the error:

(node:14316) UnhandledPromiseRejectionWarning: Error: connect
ETIMEDOUT 104.197.97.64:443
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1054:14)
(node:14316) UnhandledPromiseRejectionWarning: Unhandled promise
rejection. This error originated either by throwing inside of an async
function without a catch block, or by rejecting a promise which was
not handled with .catch(). (rejection id: 2)
(node:14316) [DEP0018] DeprecationWarning: Unhandled promise
rejections are deprecated. In the future, promise rejections that are
not handled will terminate the Node.js process with a non-zero exit
code.

I added a try catch and I still get the same ETIMEDOUT message.
